File: pubnub-gson/pubnub-gson-impl/src/integrationTest/java/com/pubnub/api/integration/LargeMessagePublishIntegrationTests.java
package com.pubnub.api.integration;

import com.pubnub.api.PubNubError;
import com.pubnub.api.PubNubException;
import com.pubnub.api.crypto.CryptoModule;
import com.pubnub.api.integration.util.BaseIntegrationTest;
import com.pubnub.api.java.PubNub;
import com.pubnub.api.java.callbacks.SubscribeCallback;
import com.pubnub.api.models.consumer.PNStatus;
import com.pubnub.api.models.consumer.history.PNFetchMessagesResult;
import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
import org.awaitility.Awaitility;
import org.awaitility.Durations;
import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Ignore;
import org.junit.Test;

import java.util.Collections;
import java.util.concurrent.atomic.AtomicBoolean;

import static com.pubnub.api.integration.util.Utils.random;
import static com.pubnub.api.integration.util.Utils.randomChannel;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@Ignore("Large message (V2 publish) integration tests. Remove @Ignore to run on demand.")
public class LargeMessagePublishIntegrationTests extends BaseIntegrationTest {

    @Override
    protected void onBefore() {

    }

    @Override
    protected void onAfter() {

    }

    @Test
    public void testPublishLargeMessageOverV1Limit_Post() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        // Create a message larger than V1 POST limit (32,768 bytes)
        // Account for JSON quotes (2 bytes), so 40KB string ensures we exceed the limit
        final String largeMessage = generateLargeString(40_000);

        pubNub.publish()
                .message(largeMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.TEN_SECONDS).untilTrue(success);
    }

    @Test
    public void testPublishLargeMessageOverV1Limit_Get() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        // Create a message larger than V1 GET path limit (32,752 bytes)
        final String largeMessage = generateLargeString(40_000);

        // Even with usePost=false, large messages should be routed to V2 POST
        pubNub.publish(largeMessage, expectedChannel)
                .usePOST(false)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.TEN_SECONDS).untilTrue(success);
    }

    @Test
    public void testPublishLargeMessage_History() throws PubNubException {
        final String expectedChannel = randomChannel();
        final String largeMessage = generateLargeString(50_000);

        pubNub.publish()
                .message(largeMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .sync();

        pause(3);

        final PNFetchMessagesResult historyResult = pubNub.fetchMessages()
                .channels(Collections.singletonList(expectedChannel))
                .maximumPerChannel(1)
                .sync();

        assertNotNull(historyResult);
        assertEquals(1, historyResult.getChannels().size());
        assertEquals(1, historyResult.getChannels().get(expectedChannel).size());
        assertEquals(largeMessage, historyResult.getChannels().get(expectedChannel).get(0).getMessage().getAsString());
    }

    @Test
    public void testPublishLargeMessage_Receive() throws PubNubException {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        final String largeMessage = generateLargeString(50_000);

        pubNub.addListener(new SubscribeCallback() {
            @Override
            public void status(@NotNull PubNub pubnub, @NotNull PNStatus status) {
                // Intentionally left empty
            }

            @Override
            public void message(@NotNull PubNub pubnub, @NotNull PNMessageResult message) {
                if (largeMessage.equals(message.getMessage().getAsString())) {
                    success.set(true);
                }
            }
        });

        subscribeToChannel(pubNub, expectedChannel);

        pubNub.publish()
                .message(largeMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .sync();

        Awaitility.await().atMost(Durations.TEN_SECONDS).untilTrue(success);
    }

    @Test
    public void testPublishLargeJsonObject_History() throws PubNubException, JSONException {
        final String expectedChannel = randomChannel();
        // Create a large JSON object (> 32KB)
        final JSONObject largePayload = new JSONObject();
        largePayload.put("id", random());
        largePayload.put("data", generateLargeString(40_000));
        largePayload.put("timestamp", System.currentTimeMillis());

        pubNub.publish()
                .message(largePayload)
                .channel(expectedChannel)
                .usePOST(true)
                .sync();

        pause(3);

        final PNFetchMessagesResult historyResult = pubNub.fetchMessages()
                .channels(Collections.singletonList(expectedChannel))
                .maximumPerChannel(1)
                .sync();

        assertNotNull(historyResult);
        assertEquals(1, historyResult.getChannels().size());
        assertEquals(1, historyResult.getChannels().get(expectedChannel).size());
        final JSONObject receivedMessage = new JSONObject(
                historyResult.getChannels().get(expectedChannel).get(0).getMessage().toString());
        assertEquals(largePayload.getString("id"), receivedMessage.getString("id"));
        assertEquals(largePayload.getString("data"), receivedMessage.getString("data"));
    }

    @Test
    public void testPublish100KBMessage() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        final String largeMessage = generateLargeString(100_000);

        pubNub.publish()
                .message(largeMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.TEN_SECONDS).untilTrue(success);
    }

    @Test
    public void testPublish500KBMessage() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        final String largeMessage = generateLargeString(500_000);

        pubNub.publish()
                .message(largeMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.ONE_MINUTE).untilTrue(success);
    }

    @Test
    public void testPublish2MBMessage() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        // We can't send exactly 2MB of raw content because the payload grows due to
        // JSON serialization overhead and encryption (if configured).
        final String largeMessage = generateLargeString(1_992_000);

        pubNub.publish(largeMessage, expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.ONE_MINUTE).untilTrue(success);
    }

    @Test
    public void testPublish2MBMessageWithCrypto() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        // We can't send exactly 2MB of raw content because the payload grows due to
        // JSON serialization overhead and encryption (if configured).
        final String largeMessage = generateLargeString(1_400_000);
        final PubNub pubnubWithCrypto = getPubNub(builder -> builder.cryptoModule(CryptoModule.createAesCbcCryptoModule("test", false)));


        pubnubWithCrypto.publish(largeMessage, expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertFalse(result.isFailure());
                    assertTrue(result.getOrNull().getTimetoken() > 0);
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.ONE_MINUTE).untilTrue(success);
    }

    @Test
    public void testPublishOversizedMessage_ShouldFail() {
        final AtomicBoolean success = new AtomicBoolean();
        final String expectedChannel = randomChannel();
        // Create a message > 2MB which should be rejected by client side validation
        final String oversizedMessage = generateLargeString(2_500_000);

        pubNub.publish()
                .message(oversizedMessage)
                .channel(expectedChannel)
                .usePOST(true)
                .async((result) -> {
                    assertTrue(result.isFailure());
                    PubNubException exception = result.exceptionOrNull();
                    assertNotNull(exception);
                    assertEquals(PubNubError.MESSAGE_TOO_LARGE, exception.getPubnubError());
                    assertEquals(PubNubError.MESSAGE_TOO_LARGE.getMessage(), exception.getErrorMessage());
                    success.set(true);
                });

        Awaitility.await().atMost(Durations.ONE_MINUTE).untilTrue(success);
    }

    private String generateLargeString(int size) {
        StringBuilder sb = new StringBuilder(size);
        for (int i = 0; i < size; i++) {
            sb.append('x');
        }
        return sb.toString();
    }
}

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/LargeMessagePublishIntegrationTests.kt
package com.pubnub.api.integration

import com.pubnub.api.PubNub
import com.pubnub.api.PubNubError
import com.pubnub.api.PubNubException
import com.pubnub.api.callbacks.SubscribeCallback
import com.pubnub.api.crypto.CryptoModule
import com.pubnub.api.models.consumer.PNBoundedPage
import com.pubnub.api.models.consumer.PNStatus
import com.pubnub.api.models.consumer.pubsub.PNMessageResult
import com.pubnub.api.v2.callbacks.getOrThrow
import com.pubnub.test.CommonUtils.randomChannel
import com.pubnub.test.CommonUtils.randomValue
import com.pubnub.test.CommonUtils.retry
import com.pubnub.test.await
import com.pubnub.test.listen
import com.pubnub.test.subscribeToBlocking
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Ignore
import org.junit.Test
import java.util.concurrent.atomic.AtomicBoolean

// make sure feature is enabled on keyset publish_v2_enabled=1 publish_v2_max_size=2097152
@Ignore("Large message (V2 publish) integration tests. Remove @Ignore to run on demand.")
class LargeMessagePublishIntegrationTests : BaseIntegrationTest() {
    @Test
    fun testPublishLargeMessageOverV1Limit_Post() {
        val expectedChannel = randomChannel()
        // Create a message larger than V1 POST limit (32,768 bytes)
        // Account for JSON quotes (2 bytes), so 40KB string ensures we exceed the limit
        val largeMessage = "x".repeat(40_000)

        pubnub.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).await { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublishLargeMessageOverV1Limit_Get() {
        val expectedChannel = randomChannel()
        // Create a message larger than V1 GET path limit (32,752 bytes)
        val largeMessage = "x".repeat(40_000)

        // Even with usePost=false, large messages should be routed to V2 POST
        server.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = false,
        ).await { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublishLargeMessage_History() {
        val expectedChannel = randomChannel()
        val largeMessage = "x".repeat(50_000)

        pubnub.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).sync()

        retry {
            pubnub.fetchMessages(
                channels = listOf(expectedChannel),
                page = PNBoundedPage(limit = 1),
            ).sync().run {
                assertEquals(1, channels.size)
                assertEquals(1, channels[expectedChannel]!!.size)
                assertEquals(largeMessage, channels[expectedChannel]!![0].message.asString)
            }
        }
    }

    @Test
    fun testPublishLargeMessage_Receive() {
        val expectedChannel = randomChannel()
        val largeMessage = "x".repeat(50_000)
        val success = AtomicBoolean()

        val observer = createPubNub {}

        pubnub.addListener(
            object : SubscribeCallback() {
                override fun status(
                    pubnub: PubNub,
                    pnStatus: PNStatus,
                ) {
                }

                override fun message(
                    pubnub: PubNub,
                    event: PNMessageResult,
                ) {
                    if (event.message.asString == largeMessage) {
                        success.set(true)
                    }
                }
            },
        )

        pubnub.subscribeToBlocking(expectedChannel)

        observer.publish(
            message = largeMessage,
            channel = expectedChannel,
            usePost = true,
        ).sync()

        success.listen()
    }

    @Test
    fun testPublishLargeJsonObject_History() {
        val expectedChannel = randomChannel()
        // Create a large JSON object (> 32KB)
        val largePayload = JSONObject().apply {
            put("id", randomValue())
            put("data", "x".repeat(40_000))
            put("timestamp", System.currentTimeMillis())
        }

        pubnub.publish(
            channel = expectedChannel,
            message = largePayload,
            usePost = true,
        ).sync()

        retry {
            pubnub.fetchMessages(
                channels = listOf(expectedChannel),
                page = PNBoundedPage(limit = 1),
            ).sync().run {
                assertEquals(1, channels.size)
                assertEquals(1, channels[expectedChannel]!!.size)
                val receivedMessage = JSONObject(channels[expectedChannel]!![0].message.toString())
                assertEquals(largePayload.getString("id"), receivedMessage.getString("id"))
                assertEquals(largePayload.getString("data"), receivedMessage.getString("data"))
            }
        }
    }

    @Test
    fun testPublish100KBMessage() {
        val expectedChannel = randomChannel()
        val largeMessage = "x".repeat(100_000)

        pubnub.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).await { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublish500KBMessage() {
        val expectedChannel = randomChannel()
        val largeMessage = "x".repeat(500_000)

        server.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).await { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublish2MBMessage() {
        val expectedChannel = randomChannel()
        // We can't send exactly 2MB of raw content because the payload grows due to
        // JSON serialization overhead and encryption (if configured).
        val largeMessage = "x".repeat((1.9 * 1024 * 1024).toInt()) // 1,992,294 chars

        pubnub.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).await(seconds = 10) { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublish2MBMessageWithCrypto() {
        val expectedChannel = randomChannel()
        val pubnubWithCrypto =
            createPubNub {
                cryptoModule = CryptoModule.createAesCbcCryptoModule("test", false)
            }

        // We can't send exactly 2MB of raw content because the payload grows due to
        // JSON serialization overhead and encryption (if configured).
        val largeMessage = "x".repeat((1.4 * 1024 * 1024).toInt())

        pubnubWithCrypto.publish(
            channel = expectedChannel,
            message = largeMessage,
            usePost = true,
        ).await(seconds = 15) { result ->
            assertFalse(result.isFailure)
            assertTrue(result.getOrThrow().timetoken > 0)
        }
    }

    @Test
    fun testPublishOversizedMessage_ShouldFail() {
        val expectedChannel = randomChannel()
        // Create a message > 2MB which should be rejected by client side validation
        val oversizedMessage = "x".repeat(2_500_000)

        pubnub.publish(
            channel = expectedChannel,
            message = oversizedMessage,
            usePost = true,
        ).await { result ->
            assertTrue(result.isFailure)
            val exception = result.exceptionOrNull()
            assertNotNull(exception)
            assertTrue(exception is PubNubException)
            assertEquals(PubNubError.MESSAGE_TOO_LARGE, (exception as PubNubException).pubnubError)
            assertEquals(PubNubError.MESSAGE_TOO_LARGE.message, exception.errorMessage)
        }
    }
}
